File: backend/app/schemas/monitoring.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class MonitoringKPITrend(BaseModel):
    total_runs_pct: float | None
    error_rate_pct: float | None
    avg_latency_pct: float | None
    p95_latency_pct: float | None


class MonitoringKPI(BaseModel):
    total_runs: int
    error_rate: float
    avg_latency_ms: float | None
    p95_latency_ms: float | None
    trend: MonitoringKPITrend


class AgentHealthRow(BaseModel):
    agent_id: uuid.UUID
    agent_name: str
    runs: int
    error_rate: float
    avg_latency_ms: float | None
    p95_latency_ms: float | None
    null_rate: float
    status: Literal["healthy", "warning", "critical", "dormant"]
    last_run_at: datetime | None


class SlowChainRow(BaseModel):
    agent_id: uuid.UUID
    agent_name: str
    chain_id: uuid.UUID | None
    chain_name: str
    p95_latency_ms: float
    calls: int


class MonitoringSummaryResponse(BaseModel):
    kpi: MonitoringKPI
    agent_health: list[AgentHealthRow]
    slow_chains: list[SlowChainRow]
    range_hours: int

File: frontend/src/types/monitoring.ts
export interface MonitoringKPI {
  total_runs: number;
  error_rate: number;       // 0.0–1.0
  avg_latency_ms: number | null;
  p95_latency_ms: number | null;
  trend: {
    total_runs_pct: number | null;
    error_rate_pct: number | null;
    avg_latency_pct: number | null;
    p95_latency_pct: number | null;
  };
}

export type AgentHealthStatus = 'healthy' | 'warning' | 'critical' | 'dormant' | 'down' | 'degraded';

export interface AgentHealthRow {
  agent_id: string;
  agent_name: string;
  runs: number;
  error_rate: number;       // 0.0–1.0
  avg_latency_ms: number | null;
  p95_latency_ms: number | null;
  null_rate: number;        // 0.0–1.0 (null output rate)
  status: AgentHealthStatus;
  last_run_at: string | null;
  availability: 'up' | 'down' | 'degraded' | null;
  availability_latency_ms: number | null;
}

export interface SlowChainRow {
  agent_id: string;
  agent_name: string;
  chain_name: string;
  chain_id?: string;
  p95_latency_ms: number;
  calls: number;
}

export interface MonitoringSummary {
  kpi: MonitoringKPI;
  agent_health: AgentHealthRow[];
  slow_chains: SlowChainRow[];
  range_hours: number;
}
